### PR TITLE
debuginfo: Improve debuginfo quality checking using metadata

### PR DESCRIPTION
**Pull request by kakkoyun**
- Add source file hash to the metadata file
- Reduce metadata states
- Check if metadata is stale
- Check and handle the file corruption
  - Using a limited buffer 

Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

Fixes https://github.com/parca-dev/parca/issues/1097

Blocked by https://github.com/parca-dev/parca/pull/1265

### CODE REVIEW

File: pkg/debuginfo/store.go
Comment (brancz) [inline pkg/debuginfo/store.go:324]: elf expects to be a very specific magic number, no? I would not just match on `'[0 0 0 0]'`, but any `bad magic number` error.

**Comment by brancz** [comment]
if we've tested this, I'm happy to merge this

**Comment by kakkoyun** [comment]
It's not fully tested. I have parked it for now, and I'll return to it soonish.

**Comment by kakkoyun** [comment]
In subsequent PRs. We should add integration tests for `debuginfo.Store.Exists` and `debuginfo.Store.Upload` using the `debuginfo.Client`.

**Comment by kakkoyun** [comment]
@javierhonduco I'll wait for your input on this before I proceed with this.

File: pkg/debuginfo/metadata.go
Comment (javierhonduco) [inline pkg/debuginfo/metadata.go:41]: The reason why this state exists is to be able to backfill the metadata files in the future, but happy to delete it if it's not needed!

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:175]: The hash is set once the file is persisted in `uploaded()` I think, so we don't have to check the state, as we know it will be `uploaded`, no?

Side note: perhaps checking for the state first can be clearer, like:

```
switch state:
  case metadataStateUploaded:
     // we know we should have a hash by now
  case metadataStateUploading:
     // we should check if the file is stale which means that it never completed uploading and the client should re-upload
...
```

What do you think?

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:240]: nit: we don't need the `metadataFile.Hash != ""` check as we will check if it's equal to the sent hash below, and it can't be an empty string

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:281]: Which kind of problems are you anticipating?

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:289]: This is beautiful, so unixy! :)

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:295]: I don't think there's a better way to do this, but I am wondering what can we do to reduce memory usage.

My concern is that we will store the debug info files we are uploading in memory which can put a lot of memory pressure in the box. By using the pool, we will make sure that we don't allocate too much, which is good! But the memory will still be used, and when several multi-gigabyte debug files are uploaded at the same time, we will have to store them in memory.

Please correct me if I am wrong and happy to hear what you think!

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:375]: This looks like a great candidate for another metadata field! We can do this in another PR, but I think it would help to debug gnarly situations, as well as potentially making the code easier to follow

**Review by javierhonduco — COMMENTED** [review]
Great stuff!! Loving this, and learned a bunch of Go-isms on the way! 

Left a couple of nits and some questions, and will test this on my machine right now!

File: proto/parca/debuginfo/v1alpha1/debuginfo.proto
Comment (kakkoyun) [reply proto/parca/debuginfo/v1alpha1/debuginfo.proto:46]: But it's not the hash of the debug information. I can work on the wording. However, it was the mistake previously we made we should store the hash of the source we extracted debug information from. I'm open to suggestions for the wording.

File: pkg/symbolizer/symbolizer.go
Comment (kakkoyun) [reply pkg/symbolizer/symbolizer.go:60]: I actually even get rid of these logs, maybe. It really creates clutter and most of the time, it's hard to determine the cause from this log line. I'll think of a better way for this.

File: proto/parca/debuginfo/v1alpha1/debuginfo.proto
Comment (javierhonduco) [reply proto/parca/debuginfo/v1alpha1/debuginfo.proto:46]: Ohhh! I see, I misunderstood what it was, I thought it was `hash(extracted_debug_file)` instead of `hash(original_executable)`. Everything makes sense now 😄

File: pkg/symbol/elfutils/elfutils.go
Comment (kakkoyun) [reply pkg/symbol/elfutils/elfutils.go:159]: I have forgotten to update the comment :see_no_evil: I had a lot of back and forth on this. None of the errors is an unrecoverable error in this context. And the existence of the error indicates if it's valid or not. That's why I went for a simpler approach. Otherwise, error handling gets super messy.

File: pkg/debuginfo/metadata.go
Comment (kakkoyun) [reply pkg/debuginfo/metadata.go:102]: And I haven't noticed it during reviews :D

File: pkg/debuginfo/metadata.go
Comment (kakkoyun) [reply pkg/debuginfo/metadata.go:103]: I'm happy to improve the design. Leaking the internal state type to the caller just feels wrong. We should just carry over the state machine logic into these methods to have the best of both worlds. But I'll leave it to consecutive PRs.

File: pkg/symbol/elfutils/elfutils.go
Comment (javierhonduco) [reply pkg/symbol/elfutils/elfutils.go:159]: Sure thing! Either way is fine :D

File: pkg/debuginfo/metadata.go
Comment (kakkoyun) [reply pkg/debuginfo/metadata.go:41]: The unknown state or lack of the metadata file implies this state. Since we don't have a lot of historical data, this shouldn't be a huge problem, don't you think?

File: pkg/debuginfo/store.go
Comment (kakkoyun) [reply pkg/debuginfo/store.go:175]: Good point. I think I've overlooked this.

File: pkg/debuginfo/metadata.go
Comment (javierhonduco) [reply pkg/debuginfo/metadata.go:108]: That's fair, it's definitely not a big deal. I was thinking of being able to inspect the files for patterns and correlations, such as trying to see if more corrupted files happen after a certain time (imagine, a new release, or if the server is overloaded, etc).

But I guess that as you said, the window for inspecting such files is small as they'll most likely be overwritten later on. 

We can always use some Prometheus counters to understand the failures and find correlations!

File: pkg/debuginfo/store.go
Comment (kakkoyun) [reply pkg/debuginfo/store.go:281]: I haven't properly checked if we reuse the byte slices underneath for these buffers. Because the wrapper buffers don't check boundaries while reading, they read until the end of the []byte underneath. This could be a problem if we got a reused []byte that is larger than the current data. I might be overthinking and missing things in here. That's why I want to double-check before merging.

File: pkg/debuginfo/metadata.go
Comment (javierhonduco) [reply pkg/debuginfo/metadata.go:41]: Thought it would be nice to have an explicit state for this situation, but it's not an issue to leave it like this :)

File: pkg/debuginfo/store.go
Comment (javierhonduco) [reply pkg/debuginfo/store.go:281]: Make sense, thanks for explaining!! I think that for this part, it would be good to have someone with actual Go expertise to take a look 😄

File: pkg/debuginfo/store.go
Comment (kakkoyun) [reply pkg/debuginfo/store.go:295]: You're definitely right. That's why this is the last piece that I have added. I really don't want to do this, but it's the only way to validate completely what we have received. If we're willing to loosen up the guarantees, we can get by just reading the first few bytes. And if something happens during symbolization, we can mark this file as corrupted (which we already do). We just need to make sure the error that we receive is not recoverable. (And this is what I meant by the last to-do item).

File: pkg/debuginfo/store.go
Comment (kakkoyun) [reply pkg/debuginfo/store.go:375]: This is a to-do item that I have on the agent side. I think we should add a source type enum file to metadata and the API. So that this could help us debug issues. And we can improve the state machine we have in hear.

Comment (kakkoyun) [reply pkg/debuginfo/store.go:375]: What do you think, shall we just add it as part of this PR?

Comment (javierhonduco) [reply pkg/debuginfo/store.go:375]: Either works, but this PR is already quite comprehensive, should we do it in another PR?

Comment (kakkoyun) [reply pkg/debuginfo/store.go:374]: Created an issue for it https://github.com/parca-dev/parca/issues/1258

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:276]: Oh, this is great to know and take into account!! 😄

File: pkg/debuginfo/store.go
Comment (javierhonduco) [inline pkg/debuginfo/store.go:292]: Nice!!

File: pkg/symbol/elfutils/elfutils.go
Comment (javierhonduco) [inline pkg/symbol/elfutils/elfutils.go:178]: This is fantastic!! Love how you are validating a bunch of things I did not even think about!! 💯

**Review by javierhonduco — APPROVED** [review]
Looks great!! 

Given that there are many changes here, could you run it for a little while on some test binary and see if everything works as expected? 

🚢

**Comment by kakkoyun** [comment]
This is still being tested and blocked by https://github.com/parca-dev/parca/pull/1265

**Review by javierhonduco — APPROVED** [review]
Amazing work!

File: pkg/debuginfo/metadata.go
Comment (brancz) [inline pkg/debuginfo/metadata.go:101]: Why is this needed? The BuildID is already in the path of the object, no?

File: pkg/debuginfo/store.go
Comment (brancz) [inline pkg/debuginfo/store.go:296]: This took me a couple of reads to understand, and I have a few questions about this.

1) Is the pooling really necessary? It seems that we'd only ever allocate max 64 bytes per upload.
2) I think for future me I would need a comment that this limitio Writer does not error when you attempt to write more than 64 bytes.
3) If I'm reading the logic right, then we're not actually validating the elf header until after we uploaded to the bucket. Isn't the whole point to prevent that?

File: pkg/debuginfo/metadata.go
Comment (javierhonduco) [reply pkg/debuginfo/metadata.go:101]: I think it would be convenient to have the build_id here, in general. Especially since it's typed and not a loose string that might change in the future and might make use add awkward build_id parsing out of the path

Comment (brancz) [reply pkg/debuginfo/metadata.go:101]: I don't feel strongly. I'm ok with keeping it.

File: pkg/debuginfo/store.go
Comment (kakkoyun) [reply pkg/debuginfo/store.go:296]: 1. Yes, pooling is not strictly necessary here. Since this is a frequent recurring task, I've put the pool there. If you think it's overkill, I'm happy t remove it.
2. I should have commented and added an explanation. I will do that.
3. Yes, this happens after the fact. We're optimistically uploading directly to the bucket and if something goes wrong we mark it as corrupted, so it could be overwritten, in subsequent calls. We only want to make sure we don't read a corrupted file while symbolizing. I wanted to prevent any form of buffering this data on the server-side thus the optimistic writes directly to the object-store. I'm open to other suggestions.

Comment (brancz) [reply pkg/debuginfo/store.go:296]: Oooooooo I see! Could you add what you wrote in 3 as a comment as well? :D 

Now it's all clear 💯 

(and yeah I think the pool is unnecessary)

**Review by brancz — APPROVED** [review]
Very nice! 💯

**Pull request merged**
